### ISSUE TITLE
tf.keras.layers.Layer.weights is not ordered in graph mode

### ISSUE THREAD

**Issue by MathiasToftas** [tensorflow/tensorflow#57438]
<details><summary>Click to expand!</summary> 
 
 ### Issue Type

Bug

### Source

binary

### Tensorflow Version

v2.9.0-18-gd8ce9f9c301 2.9.1

### Custom Code

Yes

### OS Platform and Distribution

Linux Ubuntu 20.04.4

### Python version

3.8

### Current Behaviour?

```shell
When loading a 'saved_model' (tf.saved_model.load), the 'weights' attribute found in tf.keras.layers.Layer gives the model weights in a seemingly random order. If you convert the same 'saved_model' to a tflite model (tf.lite.TFLiteConverter.from_saved_model), the weights are instead given in an expected fixed order.

Since the 'saved_model' and the tflite model are generated from the same source, one expects the same outcome when calling the 'same' method
```

### Standalone code to reproduce the issue

```shell
https://colab.research.google.com/drive/1AE0dS1vJXV3D_NL29NjkVwavcszximFY?usp=sharing
```

### Relevant log output

_No response_</details>

**Comment by tilakrayal** [comment]
@MathiasToftas,
**Layer.weights** represents the variables that parameterize the forward pass of a Layer/Model.

It contains variables that belong to any layer, model, or tf.Module attached to the layer (as well as individual Variable instances attached to it).

It does not contain variables from metrics, optimizers, callbacks, etc. because these variables do not parameterize the forward pass. It would be quite error prone to include metric variables in particular, since it would create the risk that these variables get updated during backprop, invalidating metric results (since trainable defaults to True for all variables).

**Comment by MathiasToftas** [comment]
My class is a tf.Module, and when accessing the weights attribute all of the weights and biases are returned as expected. The issue is that the list of weights is not always in the same order. When loading my model as a "saved_model" and accessing the weights, the order is seemingly set randomly every time the model is loaded, but if the same model is converted to a tflite model the order is always the same, even when reloading the model. Is this expected behavior? In my mind, the order should always be the same regardless, especially so since both the "saved_model" and tflite model are generated from the same code.

**Comment by tilakrayal** [comment]
@MathiasToftas,
This is expected since keras won't recursively visit all the attributes for a random class to find all the weights. If you would like keras to track the weights of your class, you could either make it a keras layer, or a tf.Module, which will give keras an indication that this object will contain weights. Thank you!

**Comment by MathiasToftas** [comment]
I'm not sure if I'm being unclear so just to make it crystal. My model does inherit from tf.Module, and clearly this is the case as keras does keep track of the weights and gives a complete list of all weights when the layers attribute is accessed. 

The issue is that that the order of the weights in the list is not consistent between 'environments'. 
- If the Layers attribute is accessed from the model in memory (normal usage, no loading or saving of the model in various formats), it produces the list in a consistent order. 
- If you save the model as a 'saved_model' and then load it, the weights are given in a new random order when accessing Layers. Said order is randomized again each time the model is loaded.
- If you convert the 'saved_model' to a tensorflow lite model, the Layer attribute once again gives the same consistent ordering of the weights as in the 'normal' case.

**Comment by tilakrayal** [comment]
@MathiasToftas,
Thanks for opening this issue. Development of keras moved to another [repository](https://github.com/keras-team/keras/issues). 
 
 
Could you please post this issue on keras-team/keras [repo](https://github.com/keras-team/keras/issues).
To know more please refer:
https://discuss.tensorflow.org/t/keras-project-moved-to-new-repository-in-https-github-com-keras-team-keras/1999
Thank you!
